src/wasm/literal.cpp

```diff
@@ -29,37 +29,52 @@ namespace wasm {
 
 template<int N> using LaneArray = std::array<Literal, N>;
 
+Literal::Literal(Type type) : type(type) {
+  assert(type != Type::unreachable && (!type.isRef() || type.isNullable()));
+  if (type.isException()) {
+    new (&exn) std::unique_ptr<ExceptionPackage>();
+  } else {
+    memset(&v128, 0, 16);
+  }
+}
+
 Literal::Literal(const uint8_t init[16]) : type(Type::v128) {
   memcpy(&v128, init, 16);
 }
 
 Literal::Literal(const Literal& other) : type(other.type) {
-  TODO_SINGLE_COMPOUND(type);
-  switch (type.getBasic()) {
-    case Type::i32:
-    case Type::f32:
-      i32 = other.i32;
-      break;
-    case Type::i64:
-    case Type::f64:
-      i64 = other.i64;
-      break;
-    case Type::v128:
-      memcpy(&v128, other.v128, 16);
-      break;
-    case Type::funcref:
-      func = other.func;
-      break;
-    case Type::exnref:
-      // Avoid calling the destructor on an uninitialized value
+  if (type.isException()) {
+    // Avoid calling the destructor on an uninitialized value
+    if (other.exn) {
       new (&exn) auto(std::make_unique<ExceptionPackage>(*other.exn));
-      break;
-    case Type::none:
-    case Type::nullref:
-      break;
-    case Type::externref:
-    case Type::unreachable:
-      WASM_UNREACHABLE("unexpected type");
+    } else {
+      new (&exn) std::unique_ptr<ExceptionPackage>();
+    }
+  } else if (type.isFunction()) {
+    func = other.func;
+  } else {
+    TODO_SINGLE_COMPOUND(type);
+    switch (type.getBasic()) {
+      case Type::i32:
+      case Type::f32:
+        i32 = other.i32;
+        break;
+      case Type::i64:
+      case Type::f64:
+        i64 = other.i64;
+        break;
+      case Type::v128:
+        memcpy(&v128, other.v128, 16);
+        break;
+      case Type::none:
+        break;
+      case Type::externref:
+        break; // null
```

Comment:
Can't this also fall through to the `WASM_UNREACHABLE` since it should have been handled up above?

Reply:
Previously this was

```cpp
    case Type::nullref:
      break;
```

but with the removal of `nullref` this now becomes what's here. No functional change to previously. Also note that this is neither an exception, a function nor a compound type, so not handled above.

Reply:
Sounds good. I think I misread this as `exnref` before 😅